File: ch3/clip3/boj11779/Main.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int INF = 1000000000;
        int n = sc.nextInt(), m = sc.nextInt();
        List<Edge>[] graph = new List[n + 1];
        int[] cost = new int[n + 1];
        int[] path = new int[n + 1];
        for(int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
            cost[i] = INF;
        }
        for(int i = 0; i < m; i++) {
            int s = sc.nextInt(), d = sc.nextInt(), c = sc.nextInt();
            graph[s].add(new Edge(d, c));
        }
        int start = sc.nextInt(), end = sc.nextInt();
        PriorityQueue<Edge> pq = new PriorityQueue<>((o1, o2) -> {
            return o1.cost - o2.cost;
        });
        pq.offer(new Edge(start, 0));
        cost[start] = 0;

        while(!pq.isEmpty()) {
            Edge now = pq.poll();
            if(cost[now.dist] < now.cost) continue;
            for(Edge next : graph[now.dist]) {
                if(cost[now.dist] + next.cost < cost[next.dist]) {
                    cost[next.dist] = cost[now.dist] + next.cost;
                    pq.offer(new Edge(next.dist, cost[next.dist]));
                    path[next.dist] = now.dist;
                }
            }
        }
        System.out.println(cost[end]);
        Stack<Integer> stack = new Stack<>();
        int now = end;
        while(now != 0) {
            stack.push(now);
            now = path[now];
        }
        System.out.println(stack.size());
        while(!stack.isEmpty()) {
            System.out.print(stack.pop() + " ");
        }
    }

}

class Edge {
    int dist, cost;
    public Edge(int dist, int cost) {
        this.dist = dist;
        this.cost = cost;
    }
}

File: ch3/clip3/boj11779/main.py
import sys
import heapq

INF = 1000000000

n = int(sys.stdin.readline())
m = int(sys.stdin.readline())

graph = [[] for _ in range(n + 1)]
cost = [INF] * (n + 1)
path = [0] * (n + 1)

for i in range(m):
    s, d, c = map(int, sys.stdin.readline().split())
    graph[s].append((d, c))

start, end = map(int, sys.stdin.readline().split())

pq = []
heapq.heappush(pq, (0, start))
cost[start] = 0

while pq:
    now_cost, now = heapq.heappop(pq)
    if cost[now] < now_cost:
        continue
    for next_city, next_cost in graph[now]:
        if cost[now] + next_cost < cost[next_city]:
            cost[next_city] = cost[now] + next_cost
            heapq.heappush(pq, (cost[next_city], next_city))
            path[next_city] = now

print(cost[end])

path_stack = []
now = end
while now != 0:
    path_stack.append(now)
    now = path[now]

print(len(path_stack))
print(" ".join(map(str, reversed(path_stack))))
